pkg/asset/manifests/azure/cluster.go

```diff
@@ -328,7 +298,91 @@ func GenerateClusterAssets(installConfig *installconfig.InstallConfig, clusterID
 		},
 	}, nil
 }
+func getSubnetSpec(installConfig *installconfig.InstallConfig, controlPlaneSubnet, computeSubnet string, securityGroup capz.SecurityGroup, subnets []*net.IPNet, nodeSubnetID string, infraID string) []capz.SubnetSpec {
+	// Set default control plane subnets for default installs.
+	defaultControlPlaneSubnet := capz.Subnets{
+		{
+			SubnetClassSpec: capz.SubnetClassSpec{
+				Name: controlPlaneSubnet,
+				Role: capz.SubnetControlPlane,
+				CIDRBlocks: []string{
+					subnets[0].String(),
+				},
+			},
+			SecurityGroup: securityGroup,
+		},
+	}
+	defaultComputeSubnetSpec := capz.SubnetSpec{
+		ID: nodeSubnetID,
+		SubnetClassSpec: capz.SubnetClassSpec{
+			Name: computeSubnet,
+			Role: capz.SubnetNode,
+			CIDRBlocks: []string{
+				subnets[1].String(),
+			},
+		},
+		SecurityGroup: securityGroup,
+	}
+
+	subnetSpec := []capz.SubnetSpec{}
+	hasControlPlaneSubnet := false
+	hasComputePlaneSubnet := false
 
+	// Add the user specified subnets to the spec.
+	// For single zone, alter the compute subnet to have a NATGateway and add default control plane subnet
+	// configuration.
+	for index, spec := range installConfig.Config.Azure.Subnets {
+		specGen := capz.SubnetSpec{
+			ID: "UNKNOWN",
+			SubnetClassSpec: capz.SubnetClassSpec{
+				Name: spec.Name,
+				Role: spec.Role,
+			},
+			SecurityGroup: securityGroup,
+		}
+		// The CIDR information is optional since it could be a byo subnet.
+		if len(spec.SubnetCIDR) != 0 {
+			specGen.CIDRBlocks = spec.SubnetCIDR
+		}
+		// If role is compute node and outbound type is single node, add a NAT gateway to the subnet.
+		// Only adding a NAT gateway to the first subnet.
+		if !hasComputePlaneSubnet && spec.Role == capz.SubnetNode && installConfig.Config.Azure.OutboundType == azure.NATGatewaySingleZoneOutboundType {
+			specGen.NatGateway = capz.NatGateway{
+				NatGatewayClassSpec: capz.NatGatewayClassSpec{Name: fmt.Sprintf("%s-natgw", infraID)},
+			}
+		} else if installConfig.Config.Azure.OutboundType == azure.NATGatewayMultiZoneOutboundType {
+			if spec.NatGatewayName != "" {
+				specGen.NatGateway = capz.NatGateway{
+					NatGatewayIP: capz.PublicIPSpec{
```

Comment:
Do we have to check if this is not a private cluster?

Reply:
Added a validation preventing creation of nat gateway in internal publishing strategy since NAT gateway creation is tied with a public IP.